charmhelpers/contrib/network/ovs/ovn.py

```diff
@@ -99,19 +99,27 @@ def cluster_status(target, schema=None):
             'cluster/status',
             schema or schema_map.get(target)).splitlines():
         if k and line.startswith(' '):
+            # there is no key which means this is a instance of a multi-line/
+            # multi-value item, populate the List which is already stored under
+            # the key.
             status[k].append(line.lstrip())
         elif ':' in line:
+            # this is a line with a key
             k, v = line.split(':', 1)
             k = k.lower()
             k = k.replace(' ', '_')
             if v:
+                # this is a line with both key and value
                 if k in ('cluster_id', 'server_id',):
                     v = v.replace('(', '')
                     v = v.replace(')', '')
                     status[k] = tuple(v.split())
                 else:
                     status[k] = v.lstrip()
             else:
+                # this is a line with only key which means a multi-line/
+                # multi-value item.  Store key as List which will be
+                # populated on subsequent iterations.
                 status[k] = []
     return status
```

Comment:
I can't comment at the top of the function, as there are no changes there, so this comment is about the whole `cluster_status` function.

The return value is complex.  It's a `Dict[str, ANY]`, but that means that every calling function has to dis-ambiguate and test each value from the dictionary to process it.  Yet, that information is *known* here.  Options:

1. Is this 3 separate functions?  If there are 3 different types of return values (in the dictionary), is the function trying to do to much.
2. If it's not, then can the different return types be separated into different dictionaries.  i.e. return a tuple:  `Tuple[Dict[str, str], Dict[str, List[str]], Dict[str, Tuple[str,str,...]]`  -- that could also be a NamedTuple to make accessing it more usable.
* Can the values be tagged in some way to indicate what they are.  e.g. NameTuples rather than Tuples.  classes?

I'm really concerned that this library function is essentially returning quite an opaque object that requires reading the code to work out what is being returned.

Reply:
The system provides its status using quite opaque types (both short and long identifiers, lists etc.).  I think having one function understand the tools output is valuable rather than having to query the system status three times to get the required information from the same output, even though it has to handle three types of data in one output.

I guess putting this into a named and typed data structure rather than a ``Dict`` can have value.  You would still need to read the code defining that data structure to understand it and use it, but it will undeniably be more clear.

The flip side is that it would require maintenance to the library any/every time the upstream project adds information to the output, but I guess it does make sense to handle that at the core, where there also is expert knowledge about what every item really means as opposed to letting downstream code cope with it as best it can.